File: packages/core/tests/errors.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { formatSnapError } from "../src/errors.js";

describe("error contract", () => {
  it("formats problem, cause, fix, and path when present", () => {
    const output = formatSnapError({
      code: "HEM_PARSE_FAILED",
      problem: "Could not parse Codex config.",
      cause: "TOML syntax error at line 12.",
      fix: "Run `hem scan --skip codex` or fix the TOML file.",
      path: "~/.codex/config.toml"
    });

    assert.match(output, /^HEM_PARSE_FAILED/);
    assert.match(output, /Problem: Could not parse Codex config\./);
    assert.match(output, /Cause: TOML syntax error at line 12\./);
    assert.match(output, /Fix: Run `hem scan --skip codex` or fix the TOML file\./);
    assert.match(output, /Path: ~\/\.codex\/config\.toml/);
  });
});

File: internal/hemcore/errors/errors_test.go
package errors_test

import (
	"strings"
	"testing"

	hemerrors "github.com/qyinm/hem/internal/hemcore/errors"
	"github.com/qyinm/hem/internal/hemcore/types"
)

func TestFormatSnapErrorMatchesContract(t *testing.T) {
	t.Parallel()
	path := "~/.codex/config.toml"
	out := hemerrors.FormatSnapError(types.SnapError{
		Code:    "HEM_PARSE_FAILED",
		Problem: "Could not parse Codex config.",
		Cause:   "TOML syntax error at line 12.",
		Fix:     "Run `hem scan --skip codex` or fix the TOML file.",
		Path:    &path,
	})
	if !strings.HasPrefix(out, "HEM_PARSE_FAILED") {
		t.Fatalf("prefix: %q", out)
	}
	for _, want := range []string{
		"Problem: Could not parse Codex config.",
		"Cause: TOML syntax error at line 12.",
		"Fix: Run `hem scan --skip codex` or fix the TOML file.",
		"Path: ~/.codex/config.toml",
	} {
		if !strings.Contains(out, want) {
			t.Fatalf("missing %q in %q", want, out)
		}
	}
}